### PR TITLE
Improve attrs hashability detection

### LINKED ISSUE #16550: False negative: attrs classes and hashability
Looks like non-frozen attrs classes are considered hashable by default. But they aren't.

The following snippet type-checks but explodes at runtime:

```python
from collections.abc import Hashable

from attrs import define


@define
class A:
    a: int


def my_func(c: Hashable) -> None:
    {c}


my_func(A(1))
```

(The behavior can be [customized](https://www.attrs.org/en/stable/hashing.html).)

This might be an easy one to fix. I can grab it when I have time, or it may be an easy first issue for someone?

### PR DESCRIPTION
**Pull request by Tinche**
Fixes #16550

Improve hashability detection for attrs classes.

I added a new parameter to `add_attribute_to_class`, `overwrite_existing`, since I needed it.

Frozen classes remain hashable, non-frozen default to no. This can be overriden by passing in `hash=True` or `unsafe_hash=True` (new parameter, added to stubs) to `define()`.

Inheritance-wise I think we're correct, if a non-hashable class inherits from a hashable one, it's still unhashable at runtime.

Accompanying tests.

### CODE REVIEW

**Review by ikonst — APPROVED** [review]

**Comment by Tinche** [comment]
@sobolevn Does this look good to you? 🙏

File: test-data/unit/check-plugin-attrs.test [hunk 1]
```diff
@@ -2321,3 +2321,60 @@ reveal_type(b.x)  # N: Revealed type is "builtins.int"
...
+
+[case testDefaultHashability]
+from typing import Hashable
+from attrs import define
+
+@define
+class A:
+    a: int
+
+a: Hashable = A(1)
```
Comment (sobolevn) [inline test-data/unit/check-plugin-attrs.test:2333]: I think that `reveal_type(A.__hash__)` is a better test.

Comment (Tinche) [reply test-data/unit/check-plugin-attrs.test]: Changed!

File: test-data/unit/check-plugin-attrs.test [hunk 2]
```diff
@@ -2321,3 +2321,55 @@ reveal_type(b.x)  # N: Revealed type is "builtins.int"
...
+
+[case testFrozenHashability]
+from typing import Hashable
+from attrs import frozen
+
+@frozen
+class A:
+    a: int
+
+a: Hashable = A(1)
```
Comment (sobolevn) [inline test-data/unit/check-plugin-attrs.test:2346]: The same for all other tests

Comment (Tinche) [reply test-data/unit/check-plugin-attrs.test]: Oops, done. I've also added an additional test for subclassing behavior.

**Review by sobolevn — APPROVED** [review]
Let's wait a bit for more possible input. I will merge this in  several days.

**Comment by brianmedigate** [comment]
@Tinche are you sure about the inheritance behavior? I just upgraded to mypy 1.9 and encountered this. I have a `frozen=True` attrs class which inherits from a non-frozen one, and it does seem to be hashable at runtime:

```python
In [1]: import attrs

In [2]: @attrs.define
   ...: class A:
   ...:     a: int
   ...:

In [3]: @attrs.define(frozen=True)
   ...: class B(A):
   ...:     b: int
   ...:

In [4]: hash(B(a=1, b=2))
Out[4]: -1340312973636267308

In [5]: hash(A(a=1))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[5], line 1
----> 1 hash(A(a=1))

TypeError: unhashable type: 'A'
```

this is using attrs 23.3.0

**Comment by Tinche** [comment]
@brianmedigate I tested the other direction :/

I'll look into handling this case in a new PR.

**Pull request merged**